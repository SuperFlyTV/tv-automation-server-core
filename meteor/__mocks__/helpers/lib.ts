import { TransformedCollection } from '../../lib/typings/meteor'
import * as _ from 'underscore'

/*
interface MockedCollection<T, Y extends any[]> {
	mockClear: jest.MockInstance<T, Y>['mockClear']
	mockReset: jest.MockInstance<T, Y>['mockReset']
}
*/
interface MockedCollection {
	mockClear: jest.MockInstance<void, any[]>['mockClear']
	mockReset: jest.MockInstance<void, any[]>['mockReset']
}

export function mockupCollection<A, B> (
	collection0: TransformedCollection<A, B>
) {

	const collection = collection0 as TransformedCollection<A, B> & MockedCollection

	const methodNames = [
		'allow',
		'deny',
		'find',
		'findOne',
		'insert',
		'rawCollection',
		'rawDatabase',
		'remove',
		'update',
		'upsert',
		'_ensureIndex',
		'_dropIndex'
	]
	_.each(methodNames, methodName => collection[methodName] = jest.fn(collection[methodName]))

	collection.mockClear = () => {
		_.each(methodNames, methodName => collection[methodName].mockClear())
	}
	collection.mockReset = () => {
		_.each(methodNames, methodName => collection[methodName].mockReset())
	}

	return collection
}
