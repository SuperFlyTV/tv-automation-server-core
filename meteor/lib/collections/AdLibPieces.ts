import { PieceGeneric } from './Pieces'
import { TransformedCollection } from '../typings/meteor'
import { registerCollection } from '../lib'
import { IBlueprintAdLibPiece } from '@sofie-automation/blueprints-integration'
import { createMongoCollection } from './lib'
import { RundownId } from './Rundowns'
import { PartId } from './Parts'
import { registerIndex } from '../database'

export interface AdLibPiece extends PieceGeneric, IBlueprintAdLibPiece {
	/** Rundown this AdLib belongs to */
	rundownId: RundownId

	/** Part this AdLib belongs to */
	partId?: PartId
}

export const AdLibPieces: TransformedCollection<AdLibPiece, AdLibPiece> = createMongoCollection<AdLibPiece>(
	'adLibPieces'
)
registerCollection('AdLibPieces', AdLibPieces)

registerIndex(AdLibPieces, {
	rundownId: 1,
	partId: 1,
	_rank: 1,
})
