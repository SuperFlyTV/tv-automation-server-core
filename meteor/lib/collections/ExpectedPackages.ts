import { ExpectedPackage } from '@sofie-automation/blueprints-integration'
import { TransformedCollection } from '../typings/meteor'
import { registerCollection, ProtectedString, hashObj, assertNever } from '../lib'
import { createMongoCollection } from './lib'
import { RundownId } from './Rundowns'
import { StudioId } from './Studios'
import { PieceId } from './Pieces'
import { registerIndex } from '../database'
import { AdLibActionId } from './AdLibActions'
import { BucketAdLibId } from './BucketAdlibs'
import { BucketAdLibActionId } from './BucketAdlibActions'
import { RundownBaselineAdLibActionId } from './RundownBaselineAdLibActions'
/*
 Expected Packages are created from Pieces in the rundown.
 A "Package" is a generic term for a "thing that can be played", such as media files, audio, graphics etc..
 The blueprints generate Pieces with expectedPackages on them.
 These are then picked up by a Package Manager who then tries to fullfill the expectations.
 Example: An ExpectedPackage could be a "Media file to be present on the location used by a playout device".
   The Package Manager will then copy the file to the right place.
*/

export type ExpectedPackageId = ProtectedString<'ExpectedPackageId'>

export type ExpectedPackageDB =
	| ExpectedPackageDBFromPiece
	| ExpectedPackageDBFromAdLibAction
	| ExpectedPackageDBFromBaselineAdLibAction
	| ExpectedPackageDBFromBucketAdLib
	| ExpectedPackageDBFromBucketAdLibAction

export enum ExpectedPackageDBType {
	PIECE = 'piece',
	ADLIB_ACTION = 'adlib_action',
	BASELINE_ADLIB_ACTION = 'baseline_adlib_action',
	BUCKET_ADLIB = 'bucket_adlib',
	BUCKET_ADLIB_ACTION = 'bucket_adlib_action',
}
export interface ExpectedPackageDBBase extends Omit<ExpectedPackage.Base, '_id'> {
	_id: ExpectedPackageId

	/** The studio of the Rundown of the Piece this package belongs to */
	studioId: StudioId

	/** Hash that changes whenever the content or version changes. See getContentVersionHash() */
	contentVersionHash: string

	pieceId: ProtectedString<any>
	fromPieceType: ExpectedPackageDBType
}
export interface ExpectedPackageDBFromPiece extends ExpectedPackageDBBase {
	fromPieceType: ExpectedPackageDBType.PIECE
	/** The Piece this package belongs to */
	pieceId: PieceId
	/** The rundown of the Piece this package belongs to */
	rundownId: RundownId
}

export interface ExpectedPackageDBFromAdLibAction extends ExpectedPackageDBBase {
	fromPieceType: ExpectedPackageDBType.ADLIB_ACTION
	/** The Piece this package belongs to */
	pieceId: AdLibActionId
	/** The rundown of the Piece this package belongs to */
	rundownId: RundownId
}
export interface ExpectedPackageDBFromBaselineAdLibAction extends ExpectedPackageDBBase {
	fromPieceType: ExpectedPackageDBType.BASELINE_ADLIB_ACTION
	/** The Piece this package belongs to */
	pieceId: RundownBaselineAdLibActionId
	/** The rundown of the Piece this package belongs to */
	rundownId: RundownId
}
export interface ExpectedPackageDBFromBucketAdLib extends ExpectedPackageDBBase {
	fromPieceType: ExpectedPackageDBType.BUCKET_ADLIB
	/** The Bucket adlib this package belongs to */
	pieceId: BucketAdLibId
}
export interface ExpectedPackageDBFromBucketAdLibAction extends ExpectedPackageDBBase {
	fromPieceType: ExpectedPackageDBType.BUCKET_ADLIB_ACTION
	/** The Bucket adlib-action this package belongs to */
	pieceId: BucketAdLibActionId
}
export const ExpectedPackages: TransformedCollection<
	ExpectedPackageDB,
	ExpectedPackageDB
> = createMongoCollection<ExpectedPackageDB>('expectedPackages')
registerCollection('ExpectedPackages', ExpectedPackages)

registerIndex(ExpectedPackages, {
	studioId: 1,
})
export function getContentVersionHash(expectedPackage: Omit<ExpectedPackage.Any, '_id'>): string {
	return hashObj({
		content: expectedPackage.content,
		version: expectedPackage.version,
		// todo: should expectedPackage.sources.containerId be here as well?
	})
}
export function getPreviewPackageSettings(
	expectedPackage: ExpectedPackage.Any
): ExpectedPackage.SideEffectPreviewSettings | undefined {
	let packagePath: string | undefined

	if (expectedPackage.type === ExpectedPackage.PackageType.MEDIA_FILE) {
		packagePath = expectedPackage.content.filePath
	} else if (expectedPackage.type === ExpectedPackage.PackageType.QUANTEL_CLIP) {
		packagePath = expectedPackage.content.guid || expectedPackage.content.title
	} else {
		assertNever(expectedPackage)
	}
	if (packagePath) {
		return {
			path: packagePath + '_preview.webm',
		}
	}
	return undefined
}
export function getThumbnailPackageSettings(
	expectedPackage: ExpectedPackage.Any
): ExpectedPackage.SideEffectThumbnailSettings | undefined {
	let packagePath: string | undefined

	if (expectedPackage.type === ExpectedPackage.PackageType.MEDIA_FILE) {
		packagePath = expectedPackage.content.filePath
	} else if (expectedPackage.type === ExpectedPackage.PackageType.QUANTEL_CLIP) {
		packagePath = expectedPackage.content.guid || expectedPackage.content.title
	} else {
		assertNever(expectedPackage)
	}
	if (packagePath) {
		return {
			path: packagePath + '_thumbnail.png',
		}
	}
	return undefined
}
