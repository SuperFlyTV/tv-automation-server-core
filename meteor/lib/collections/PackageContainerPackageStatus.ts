import { TransformedCollection } from '../typings/meteor'
import { registerCollection, ProtectedString, Time, protectString } from '../lib'
import { createMongoCollection } from './lib'
import { StudioId } from './Studios'
import { registerIndex } from '../database'
import { ExpectedPackageStatusAPI } from '@sofie-automation/blueprints-integration'
import { ExpectedPackageId } from './ExpectedPackages'

/**
 * The PackageContainerPackageStatuses-collection contains statuses about "a Package on a specific PackageContainer"
 * PackageContainerPackageStatuses are populated by the Package Manager-device and can be used to look up whether a Package
 * (that originally is specified in the ExpectedPackages collection) is present on a certain Package Container.
 *
 * Note: A "Package Container" is a generic term for "something that contains packages".
 * One  example of this could be a Media-folder (the "package container") which contains Media-files ("packages").
 */

/** Id of a package container */
export type PackageContainerId = ProtectedString<'PackageContainerId'>

export interface PackageContainerPackageStatusDB {
	_id: PackageContainerId // unique id, see getPackageContainerPackageId()

	/** The studio this PackageContainer is defined in */
	studioId: StudioId

	/** The PackageContainer the package is in */
	containerId: string

	/** The Package this status is for */
	packageId: string

	/** The status of the Package */
	status: ExpectedPackageStatusAPI.PackageContainerPackageStatus

	modified: Time
}

export const PackageContainerPackageStatuses: TransformedCollection<
	PackageContainerPackageStatusDB,
	PackageContainerPackageStatusDB
> = createMongoCollection<PackageContainerPackageStatusDB>('packageContainerStatuses')
registerCollection('PackageContainerStatuses', PackageContainerPackageStatuses)

registerIndex(PackageContainerPackageStatuses, {
	studioId: 1,
	containerId: 1,
	packageId: 1,
})

export function getPackageContainerPackageId(
	studioId: StudioId,
	containerId: string,
	packageId: string | ExpectedPackageId
): PackageContainerId {
	return protectString(`${studioId}_${containerId}_${packageId}`)
}
