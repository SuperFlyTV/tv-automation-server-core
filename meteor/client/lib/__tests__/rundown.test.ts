import { testInFiber } from '../../../__mocks__/helpers/jest'
import {
	setupDefaultStudioEnvironment,
	DefaultEnvironment,
	setupDefaultRundownPlaylist,
} from '../../../__mocks__/helpers/database'
import { RundownUtils } from '../rundown'
import { RundownPlaylists, RundownPlaylistId } from '../../../lib/collections/RundownPlaylists'

describe('client/lib/rundown', () => {
	let env: DefaultEnvironment
	let playlistId: RundownPlaylistId
	beforeAll(() => {
		env = setupDefaultStudioEnvironment()
		playlistId = setupDefaultRundownPlaylist(env).playlistId
	})
	testInFiber('RundownUtils.getResolvedSegment', () => {
		const showStyleBase = env.showStyleBase
		const playlist = RundownPlaylists.findOne(playlistId)
		if (!playlist) throw new Error('Rundown not found')

		const { currentPartInstance, nextPartInstance } = playlist.getSelectedPartInstances()

		const segments = playlist.getSegments()
		const parts = playlist.getAllOrderedParts()
		const segment = segments[0]
		const nextSegment = segments[1]

		const resolvedSegment = RundownUtils.getResolvedSegment(
			showStyleBase,
			playlist,
			segment,
			new Set(segments.slice(0, 0).map((segment) => segment._id)),
			parts.map((part) => part._id),
			currentPartInstance,
			nextPartInstance
		)
		expect(resolvedSegment).toBeTruthy()
		expect(resolvedSegment.parts).toHaveLength(2)
		expect(resolvedSegment).toMatchObject({
			// segmentExtended: SegmentExtended,
			// parts: Array<PartExtended>,
			isLiveSegment: false,
			isNextSegment: false,
			currentLivePart: undefined,
			currentNextPart: undefined,
			hasRemoteItems: false,
			hasGuestItems: false,
			hasAlreadyPlayed: false,
			autoNextPart: false,
		})
		expect(resolvedSegment).toMatchSnapshot()
	})
})
