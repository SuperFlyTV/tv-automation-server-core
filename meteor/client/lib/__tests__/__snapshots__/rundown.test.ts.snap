// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client/lib/rundown RundownUtils.getResolvedSegment 1`] = `
Object {
  "autoNextPart": false,
  "currentLivePart": undefined,
  "currentNextPart": undefined,
  "hasAlreadyPlayed": false,
  "hasGuestItems": false,
  "hasRemoteItems": false,
  "isLiveSegment": false,
  "isNextSegment": false,
  "parts": Array [
    Object {
      "instance": PartInstance {
        "_id": "randomId9012_part0_0_tmp_instance",
        "isTemporary": true,
        "part": Part {
          "_id": "randomId9012_part0_0",
          "_rank": 0,
          "duration": 20,
          "externalId": "MOCK_PART_0_0",
          "rundownId": "randomId9012",
          "segmentId": "randomId9012_segment0",
          "title": "Part 0 0",
        },
        "rundownId": "randomId9012",
        "segmentId": "randomId9012_segment0",
        "takeCount": -1,
      },
      "partId": "randomId9012_part0_0",
      "pieces": Array [
        Object {
          "instance": Object {
            "_id": "randomId9012_piece000_tmp_instance",
            "isTemporary": true,
            "partInstanceId": "randomId9012_part0_0_tmp_instance",
            "piece": Object {
              "_id": "randomId9012_piece000",
              "enable": Object {
                "start": 0,
              },
              "externalId": "MOCK_PIECE_000",
              "name": "Piece 000",
              "outputLayerId": "pgm",
              "partId": "randomId9012_part0_0",
              "rundownId": "randomId9012",
              "sourceLayerId": "cam0",
              "status": 0,
            },
            "rundownId": "randomId9012",
          },
          "maxLabelWidth": 0,
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  [Circular],
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece010_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece010",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_010",
                        "name": "Piece 010",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "cam0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece001_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece001",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_001",
                        "name": "Piece 001",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "vt0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "renderedDuration": null,
          "renderedInPoint": 0,
          "sourceLayer": Object {
            "_id": "cam0",
            "_rank": 0,
            "exclusiveGroup": "main",
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              [Circular],
              Object {
                "instance": Object {
                  "_id": "randomId9012_piece010_tmp_instance",
                  "isTemporary": true,
                  "partInstanceId": "randomId9012_part0_0_tmp_instance",
                  "piece": Object {
                    "_id": "randomId9012_piece010",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_010",
                    "name": "Piece 010",
                    "outputLayerId": "pgm",
                    "partId": "randomId9012_part0_0",
                    "rundownId": "randomId9012",
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  "rundownId": "randomId9012",
                },
                "outputLayer": Object {
                  "_id": "pgm",
                  "_rank": 0,
                  "isPGM": true,
                  "name": "PGM",
                  "sourceLayers": Array [
                    [Circular],
                    Object {
                      "_id": "vt0",
                      "_rank": 1,
                      "exclusiveGroup": "main",
                      "followingItems": Array [],
                      "name": "VT",
                      "pieces": Array [
                        Object {
                          "instance": Object {
                            "_id": "randomId9012_piece001_tmp_instance",
                            "isTemporary": true,
                            "partInstanceId": "randomId9012_part0_0_tmp_instance",
                            "piece": Object {
                              "_id": "randomId9012_piece001",
                              "enable": Object {
                                "start": 0,
                              },
                              "externalId": "MOCK_PIECE_001",
                              "name": "Piece 001",
                              "outputLayerId": "pgm",
                              "partId": "randomId9012_part0_0",
                              "rundownId": "randomId9012",
                              "sourceLayerId": "vt0",
                              "status": 0,
                            },
                            "rundownId": "randomId9012",
                          },
                          "outputLayer": [Circular],
                          "renderedDuration": null,
                          "renderedInPoint": 0,
                          "sourceLayer": [Circular],
                        },
                      ],
                      "type": 2,
                    },
                  ],
                  "used": true,
                },
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
            ],
            "type": 1,
          },
        },
        Object {
          "instance": Object {
            "_id": "randomId9012_piece001_tmp_instance",
            "isTemporary": true,
            "partInstanceId": "randomId9012_part0_0_tmp_instance",
            "piece": Object {
              "_id": "randomId9012_piece001",
              "enable": Object {
                "start": 0,
              },
              "externalId": "MOCK_PIECE_001",
              "name": "Piece 001",
              "outputLayerId": "pgm",
              "partId": "randomId9012_part0_0",
              "rundownId": "randomId9012",
              "sourceLayerId": "vt0",
              "status": 0,
            },
            "rundownId": "randomId9012",
          },
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece000_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece000",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_000",
                        "name": "Piece 000",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "cam0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "maxLabelWidth": 0,
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece010_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece010",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_010",
                        "name": "Piece 010",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "cam0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  [Circular],
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "renderedDuration": null,
          "renderedInPoint": 0,
          "sourceLayer": Object {
            "_id": "vt0",
            "_rank": 1,
            "exclusiveGroup": "main",
            "followingItems": Array [],
            "name": "VT",
            "pieces": Array [
              [Circular],
            ],
            "type": 2,
          },
        },
        Object {
          "instance": Object {
            "_id": "randomId9012_piece010_tmp_instance",
            "isTemporary": true,
            "partInstanceId": "randomId9012_part0_0_tmp_instance",
            "piece": Object {
              "_id": "randomId9012_piece010",
              "enable": Object {
                "start": 0,
              },
              "externalId": "MOCK_PIECE_010",
              "name": "Piece 010",
              "outputLayerId": "pgm",
              "partId": "randomId9012_part0_0",
              "rundownId": "randomId9012",
              "sourceLayerId": "cam0",
              "status": 0,
            },
            "rundownId": "randomId9012",
          },
          "outputLayer": Object {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": Array [
              Object {
                "_id": "cam0",
                "_rank": 0,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "Camera",
                "pieces": Array [
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece000_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece000",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_000",
                        "name": "Piece 000",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "cam0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "maxLabelWidth": 0,
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                  [Circular],
                ],
                "type": 1,
              },
              Object {
                "_id": "vt0",
                "_rank": 1,
                "exclusiveGroup": "main",
                "followingItems": Array [],
                "name": "VT",
                "pieces": Array [
                  Object {
                    "instance": Object {
                      "_id": "randomId9012_piece001_tmp_instance",
                      "isTemporary": true,
                      "partInstanceId": "randomId9012_part0_0_tmp_instance",
                      "piece": Object {
                        "_id": "randomId9012_piece001",
                        "enable": Object {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_001",
                        "name": "Piece 001",
                        "outputLayerId": "pgm",
                        "partId": "randomId9012_part0_0",
                        "rundownId": "randomId9012",
                        "sourceLayerId": "vt0",
                        "status": 0,
                      },
                      "rundownId": "randomId9012",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 2,
              },
            ],
            "used": true,
          },
          "renderedDuration": null,
          "renderedInPoint": 0,
          "sourceLayer": Object {
            "_id": "cam0",
            "_rank": 0,
            "exclusiveGroup": "main",
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              Object {
                "instance": Object {
                  "_id": "randomId9012_piece000_tmp_instance",
                  "isTemporary": true,
                  "partInstanceId": "randomId9012_part0_0_tmp_instance",
                  "piece": Object {
                    "_id": "randomId9012_piece000",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_000",
                    "name": "Piece 000",
                    "outputLayerId": "pgm",
                    "partId": "randomId9012_part0_0",
                    "rundownId": "randomId9012",
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  "rundownId": "randomId9012",
                },
                "maxLabelWidth": 0,
                "outputLayer": Object {
                  "_id": "pgm",
                  "_rank": 0,
                  "isPGM": true,
                  "name": "PGM",
                  "sourceLayers": Array [
                    [Circular],
                    Object {
                      "_id": "vt0",
                      "_rank": 1,
                      "exclusiveGroup": "main",
                      "followingItems": Array [],
                      "name": "VT",
                      "pieces": Array [
                        Object {
                          "instance": Object {
                            "_id": "randomId9012_piece001_tmp_instance",
                            "isTemporary": true,
                            "partInstanceId": "randomId9012_part0_0_tmp_instance",
                            "piece": Object {
                              "_id": "randomId9012_piece001",
                              "enable": Object {
                                "start": 0,
                              },
                              "externalId": "MOCK_PIECE_001",
                              "name": "Piece 001",
                              "outputLayerId": "pgm",
                              "partId": "randomId9012_part0_0",
                              "rundownId": "randomId9012",
                              "sourceLayerId": "vt0",
                              "status": 0,
                            },
                            "rundownId": "randomId9012",
                          },
                          "outputLayer": [Circular],
                          "renderedDuration": null,
                          "renderedInPoint": 0,
                          "sourceLayer": [Circular],
                        },
                      ],
                      "type": 2,
                    },
                  ],
                  "used": true,
                },
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
              [Circular],
            ],
            "type": 1,
          },
        },
      ],
      "renderedDuration": 3000,
      "startsAt": 0,
      "willProbablyAutoNext": false,
    },
    Object {
      "instance": PartInstance {
        "_id": "randomId9012_part0_1_tmp_instance",
        "isTemporary": true,
        "part": Part {
          "_id": "randomId9012_part0_1",
          "_rank": 1,
          "externalId": "MOCK_PART_0_1",
          "rundownId": "randomId9012",
          "segmentId": "randomId9012_segment0",
          "title": "Part 0 1",
        },
        "rundownId": "randomId9012",
        "segmentId": "randomId9012_segment0",
        "takeCount": -1,
      },
      "partId": "randomId9012_part0_1",
      "pieces": Array [],
      "renderedDuration": 3000,
      "startsAt": 3000,
      "willProbablyAutoNext": false,
    },
  ],
  "segmentExtended": Object {
    "_id": "randomId9012_segment0",
    "_rank": 0,
    "externalId": "MOCK_SEGMENT_0",
    "externalModified": 1,
    "name": "Segment 0",
    "outputLayers": Object {
      "pgm": Object {
        "_id": "pgm",
        "_rank": 0,
        "isPGM": true,
        "name": "PGM",
        "sourceLayers": Array [
          Object {
            "_id": "cam0",
            "_rank": 0,
            "exclusiveGroup": "main",
            "followingItems": Array [],
            "name": "Camera",
            "pieces": Array [
              Object {
                "instance": Object {
                  "_id": "randomId9012_piece000_tmp_instance",
                  "isTemporary": true,
                  "partInstanceId": "randomId9012_part0_0_tmp_instance",
                  "piece": Object {
                    "_id": "randomId9012_piece000",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_000",
                    "name": "Piece 000",
                    "outputLayerId": "pgm",
                    "partId": "randomId9012_part0_0",
                    "rundownId": "randomId9012",
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  "rundownId": "randomId9012",
                },
                "maxLabelWidth": 0,
                "outputLayer": [Circular],
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
              Object {
                "instance": Object {
                  "_id": "randomId9012_piece010_tmp_instance",
                  "isTemporary": true,
                  "partInstanceId": "randomId9012_part0_0_tmp_instance",
                  "piece": Object {
                    "_id": "randomId9012_piece010",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_010",
                    "name": "Piece 010",
                    "outputLayerId": "pgm",
                    "partId": "randomId9012_part0_0",
                    "rundownId": "randomId9012",
                    "sourceLayerId": "cam0",
                    "status": 0,
                  },
                  "rundownId": "randomId9012",
                },
                "outputLayer": [Circular],
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
            ],
            "type": 1,
          },
          Object {
            "_id": "vt0",
            "_rank": 1,
            "exclusiveGroup": "main",
            "followingItems": Array [],
            "name": "VT",
            "pieces": Array [
              Object {
                "instance": Object {
                  "_id": "randomId9012_piece001_tmp_instance",
                  "isTemporary": true,
                  "partInstanceId": "randomId9012_part0_0_tmp_instance",
                  "piece": Object {
                    "_id": "randomId9012_piece001",
                    "enable": Object {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_001",
                    "name": "Piece 001",
                    "outputLayerId": "pgm",
                    "partId": "randomId9012_part0_0",
                    "rundownId": "randomId9012",
                    "sourceLayerId": "vt0",
                    "status": 0,
                  },
                  "rundownId": "randomId9012",
                },
                "outputLayer": [Circular],
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
            ],
            "type": 2,
          },
        ],
        "used": true,
      },
    },
    "rundownId": "randomId9012",
    "sourceLayers": Object {
      "cam0": Object {
        "_id": "cam0",
        "_rank": 0,
        "exclusiveGroup": "main",
        "followingItems": Array [],
        "name": "Camera",
        "pieces": Array [],
        "type": 1,
      },
      "vt0": Object {
        "_id": "vt0",
        "_rank": 1,
        "exclusiveGroup": "main",
        "followingItems": Array [],
        "name": "VT",
        "pieces": Array [],
        "type": 2,
      },
    },
  },
}
`;
